### PR TITLE
Device post stop hook

### PR DESCRIPTION
**Pull request by tomponline**
This modifies the `device` interface to allow the `Stop()` function to return a `RunConfig` struct, similar to the `Start()` function. 

This allows the `Stop()` function to return a partially populated `RunConfig` struct to communicate with LXD which devices it should detach from the container and how to do it (such as what interface name to use when detaching a NIC back onto the host).

It also allows a slice of functions to be returned in the `PostHooks` variable, which LXD will then execute after the device has been detached.

This also changes when the `Stop()` function is called, it is now executed *before* any devices are detached from a running container, however the `PostHooks` are called after the detaching has been completed.

It also has a nice side-effect of mirroring the behaviour of the `Start()` function.

The reason for this change is as follows:

1. Allows NIC devices to indicate to LXD that it does not need to detach an interface (such as bridge and p2p veth devices), as the PostHooks will delete the parent device which will delete the veth peer.
2. Allows NIC devices to specify the detach hostname directly, without having to use volatile, which pushes more detach logic down into the device package.
3. Allows future unix devices to figure out which paths need to be detached inside the device package, without having to put device-type specific logic into LXD. 

This PR also adds some more tests to check for volatile key clean up.

### CODE REVIEW

File: test/suites/container_devices_nic_ipvlan.sh [hunk 1]
```diff
@@ -7,67 +7,72 @@ test_container_devices_nic_ipvlan() {
     return
   fi
 
-  ct_name="nt$$"
+  ctName="nt$$"
```
Comment (stgraber) [inline test/suites/container_devices_nic_ipvlan.sh]: Can you split that rename into its own commit so it's a bit easier to handle cherry-picks?

File: test/suites/container_devices_nic_macvlan.sh [hunk 2]
```diff
@@ -2,86 +2,94 @@ test_container_devices_nic_macvlan() {
   ensure_import_testimage
   ensure_has_localhost_remote "${LXD_ADDR}"
 
-  ct_name="nt$$"
+  ctName="nt$$"
```
Comment (stgraber) [inline test/suites/container_devices_nic_macvlan.sh]: This one too

**Comment by tomponline** [comment]
@stgraber those commits have been split.

**Comment by stgraber** [comment]
jenkins: test this please
